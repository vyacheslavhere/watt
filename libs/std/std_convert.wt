/*
 Библиотека convert.
 | имплементация to_int, to_float, to_long,
 | to_string, to_bool
 | находится в
 | > crate::vm::natives::libs::natives_convert
 */
unit convert {
    /*
     Конвертирует
     значение в int
    */
    native to_int -> 'convert@to_int'

    /*
     Конвертирует
     значение в float
    */
    native to_float -> 'convert@to_float'

    /*
     Конвертирует
     значение в string
    */
    native to_string -> 'convert@to_string'

    /*
     Конвертирует
     значение в bool
    */
    native to_bool -> 'convert@to_bool'
}