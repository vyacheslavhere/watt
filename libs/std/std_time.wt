/*
 Timestamp
*/
type Timestamp(raw) {
    // нативные функции
    native __mills -> 'timestamp@mills'
    native __month -> 'timestamp@month'
    native __year -> 'timestamp@year'
    native __day -> 'timestamp@day'
    native __hour -> 'timestamp@hour'
    native __minute -> 'timestamp@minute'
    native __second -> 'timestamp@second'

    // в строку
    fun to_string() {
        return __day(raw) + '.' + __month(raw) + '.' + __year(raw) + ' ' +
               __hour(raw) + ':' + __minute(raw) + ':' + __second(raw)
    }
}

/*
 Библиотека time.
 | имплементация now, from, format, at
 | находится в
 | > crate::vm::natives::libs::natives_time
 */
unit time {
    /*
     Текущее время
    */
    native __now -> 'time@now'

    /*
     Получение текущего времени
    */
    fun now {
        raw := self.__now()
        return new Timestamp(raw)
    }
}