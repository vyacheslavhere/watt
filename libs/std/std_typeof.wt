/*
 Библиотека typeof.
 | имплементация of, f_of,
 | находится в
 | > crate::vm::natives::libs::natives_typeof
 */

/*
 Возвращает строку с названием типа значения.
*/
native typeof -> 'typeof@of'

/*
 Возвращает строку с полным названием типа значения.
*/
native full_typeof -> 'typeof@fof'