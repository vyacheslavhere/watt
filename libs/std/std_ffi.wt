import 'std.errors'

type Library(raw_ptr) {
    fun call(name, args) {
        res := ffi.__internal_call(self.raw_ptr, name, args)

        return result.ok(res)
    }
}

unit ffi {
    native __internal_load -> 'ffi@load'
    native __internal_call -> 'ffi@call'

    fun load(path) {
        lib := __internal_load(path)

        return result.ok(new Library(lib))
    }
}