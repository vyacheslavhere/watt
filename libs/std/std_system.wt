/*
 Библиотека system.
 | имплементация находится в
 | > crate::vm::natives::libs::natives_system
 */

import 'std.errors'

// Объект системы
type System {
    // Количество ОЗУ, установленной на устройстве
    fun memory_total() {
        return system.__internal_memory_total()
    }

    // Количество использованной ОЗУ
    fun memory_used() {
        return system.__internal_memory_used()
    }

    // Количество вычислительных ядер
    fun cpu_count() {
        return system.__internal_cpu_count()
    }
}

// Объект, содержащий информацию о текущем процессе
type ThisProcess {
    // PID
    fun id() {
        return system.__internal_this_process_id()
    }

    // Выход с кодом
    fun exit(code) {
        system.__internal_process_exit(code)
    }
}

// Объект, содержащий информацию о запущенном процессе
type Process(raw_handle) {
    // Завершить процесс
    fun terminate {
        system.__internal_process_terminate(self.raw_handle)
    }

    // Ждать окончания выполнения процесса
    fun wait {
        return system.__internal_process_wait(self.raw_handle)
    }

    // Получает PID запущенного процесса
    fun id {
        return system.__internal_process_id(self.raw_handle)
    }
}

// Тип-помощник. Позволяет собирать будущий процесс.
type ProcessBuilder {
    fun spawn_with_shell(command) {
        internal := system.__internal_process_spawn_shell(command)

        return new Process(internal)
    }
}

unit system {
    native __internal_getenv -> 'system@getenv'
    native __internal_setenv -> 'system@setenv'
    native __internal_getcwd -> 'system@getcwd'
    native __internal_getargs -> 'system@getargs'
    native __internal_get_osname -> 'system@get_osname'

    native __internal_memory_total -> 'system@memory_total'
    native __internal_memory_used -> 'system@memory_used'
    native __internal_cpu_count -> 'system@cpu_count'

    native __internal_process_spawn_shell -> 'system@process_spawn_shell'
    native __internal_process_id -> 'system@process_id'
    native __internal_process_wait -> 'system@process_wait'
    native __internal_process_terminate -> 'system@process_terminate'
    
    native __internal_this_process_id -> 'system@this_process_id'
    native __internal_process_exit -> 'system@process_exit'

    // Получает переменную среды по имени
    fun env(name) {
        internal := __internal_getenv(name)

        if internal == null {
            return result.err(0)
        }

        return result.ok(internal)
    }

    // Устанавливает переменную среды
    fun set_env(name, value) {
        __internal_setenv(name, value)
    }

    // Получает текущую рабочую директорию
    fun cwd() {
        internal := __internal_getcwd()

        if internal == null {
            return result.err(0)
        }

        return result.ok(internal)
    }

    // Получает список аргументов, переданные программе
    fun get_args() {
        internal_list := __internal_getargs()

        real_list := new List()
        real_list.internal = internal_list

        return real_list
    }

    // Создает объект системы для получения данных о системе
    fun summary() {
        return new System()
    }

    // Получает текущий процесс
    fun this() {
        return new ThisProcess()
    }

    // Запуск нового процесса по команде для командной оболочки (`sh` для Unix, `cmd` для Windows)
    // TODO: Должно быть во вложенном юните `process`
    fun spawn_with_shell(command) {
        proc := new ProcessBuilder()

        return proc.spawn_with_shell(command)
    }

    fun os_name() {
        return __internal_get_osname()
    }
}