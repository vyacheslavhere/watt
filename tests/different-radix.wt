import 'std.io'

// 65535
hexa := 0xFFFF

// 4095
octa := 0o7777

// 2025
bin := 0b11111101001

io.println(hexa)
io.println(octa)
io.println(bin)
