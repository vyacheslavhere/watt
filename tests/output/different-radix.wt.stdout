65535
4095
2025
